cache/manager.go

```diff
@@ -740,6 +761,122 @@ func (cm *cacheManager) Merge(ctx context.Context, inputParents []ImmutableRef,
 	return rec.ref(true, dhs), nil
 }
 
+func (cm *cacheManager) Diff(ctx context.Context, lower, upper ImmutableRef, opts ...RefOption) (ir ImmutableRef, rerr error) {
```

Comment:
I'm not convinced in the "single layer only" diff description in the commit message. To me diff should always be the opposite step of merge. Merge appends layer blobs and diff splits them. Am I missing something?

Reply:
> with diff1 only containing a creation of of /b and no deletion

You have me confused now. You are not saying that your diff definition really `Merge(base, Diff(layer~1, layer)`, right? If so then I think this is problematic. LLB should not expose any concept of a "layer". How many actual layers map to a llb state should be defined by the storage implementation. Client can not track that and it can be different for different storage backend. Eg. in git like storage there is no concept of layers, just independent trees.

In your example I don't understand why these cases should be different. If you want a diff that only contains last layer you would do `diff(layer2, layer3)`.

If you just mean that you would like to have a "flatten" step then in that case `COPY` should be used to squash things together.

Having the diff be "proper unmerge" seems essential for it to have the similar lazy semantics as merge does. Eg. pick last layers from an image and put them on top of another base image - all without pulling down either of the images.

> llb.Copy to get this behavior, but that is slower than Diff which uses hardlinks and is lazy.

Afaics there is no reason why copy couldn't do hardlinks (or file_range) where supported as well if we just added that capability, and we should. As said previously I think this breaks the point of the lazy optimizations - skip mounts and work directly with blobs. But if you flatten then you can't work with blobs.

> If you want to "squash" multiple layers into a single one, you can do so with Diff as it's defined in this PR

But this is not what diff is made for.

Reply:
> You are not saying that your diff definition really Merge(base, Diff(layer~1, layer), right? 

No, in the PR, Diff is always just a single layer that represents the diff between the provided lower and upper, that's it.

> LLB should not expose any concept of a "layer". How many actual layers map to a llb state should be defined by the storage implementation. 

I am using the term layer out of convenience here, but we aren't actually exposing the concept of underlying layer storage to LLB. Diffs are between the results of LLB ops as far as clients are concerned, just like all the other LLB ops use results from other ops as inputs. Whether results are stored as layers and how/whether they are re-used continues to exist only in the server-side buildkitd as an implementation detail.

> In your example I don't understand why these cases should be different. If you want a diff that only contains last layer you would do diff(layer2, layer3).

> Having the diff be "proper unmerge" seems essential for it to have the similar lazy semantics as merge does. Eg. pick last layers from an image and put them on top of another base image - all without pulling down either of the images.

We don't want a diff that only includes the last layer, the point is that running the differ between snapshots A and B (call that a "simple diff") is not necessarily the same as joining together the layers separating A and B (call that unmerge). When umerge and there's >1 layer separating them, you can end up with intermediate deletes that aren't present with the simple diff. These deletes can have an impact when used as a merge input. 

There are use cases for both behaviors (simple diff and unmerge), so what I care about is finding a way to support both. It turns out, you can support both by just supporting simple diff. If you have that ability, unmerge is implemented entirely client-side by just merging together the simple diffs of the results separating lower and upper, as shown in `diff2` in the previous example. As I said, we could provide a client-side util function to help with this if desired.

> Having the diff be "proper unmerge" seems essential for it to have the similar lazy semantics as merge does. Eg. pick last layers from an image and put them on top of another base image - all without pulling down either of the images.

> But this is not what diff is made for.

Yes, unmerge will maximize blob re-use, but the simple diff approach also sometimes have advantages. That's why I want a straightforward way to support both. The efficient "squash" is a nice side effect of the approach, it's not meant to be the main point.

> Afaics there is no reason why copy couldn't do hardlinks (or file_range) where supported as well if we just added that capability, and we should. As said previously I think this breaks the point of the lazy optimizations - skip mounts and work directly with blobs. But if you flatten then you can't work with blobs.

Unless copy becomes capable of retaining deletes when flattening, it wouldn't be quite the same as Diff, but we could figure that out later.

Overall, all I care about is being able to support both "simple diff" and "unmerge" behaviors in the cases where either could be used and have different results. I personally like the approach in the PR best because it's simple but it doesn't really matter much if we reach the same feature set a different way. LLB is meant to be abstracted over anyways.

How about this: I update this PR to have the unmerge behavior, which will be the default (it's a very easy update actually). Then, in a later follow-up we can consider either 1) adding a flag to DiffOp that enables the "simple diff" behavior to be forced on, or 2) find a way to update FileOp.Copy to have equivalent behavior+performance for that use case? That seems like it should unblock your concern for this PR while not ruling out the rest of our options in the future.

Reply:
> How about this: I update this PR to have the unmerge behavior, which will be the default (it's a very easy update actually). Then, in a later follow-up we can consider either 1) adding a flag to DiffOp that enables the "simple diff" behavior to be forced on, or 2) find a way to update FileOp.Copy to have equivalent behavior+performance for that use case? That seems like it should unblock your concern for this PR while not ruling out the rest of our options in the future.

Ok. If needed fileop.copy can even use a different internal function when it detects that source is a diff-ref and whole root is copied. I think it is much clearer if both merge&diff have completely lazy logic, meaning they can 100% work with remote blobs only. And copy is the operation that always requires local data, the same requirements that are always needed for "flatten" behavior.

Reply:
I'm not completely against "flatten" opt-in method as well if there are good reasons why it might be better. But the user needs to understand that it comes with quite serious downsides for layer reuse. Most current "let's flatten as optimization" advice you see in current Dockerfiles is wrong.

Reply:
Updated with this behavior. I implemented it by updating `cacheManager.Diff` to check if lower is an ancestor of upper and, if so, create a merge ref of the single layer diffs between them. I like this approach because it results in the desired optimal behavior while keeping the extra complication isolated to just the Diff method and ensuring that any future optimizations to Merge (such as de-duplicating equivalent merges) will have the potential to benefit this implementation too. Additionally, if we decide to eventually support forcing the simple diff behavior on, it will be as simple as adding a flag that lets us skip that section of code rather than updating things all over the place.

> I'm not completely against "flatten" opt-in method as well if there are good reasons why it might be better

The flattening behavior is more of a side-effect, not the main point. The main point is that the actual contents of the simple diff and the unmerge can be different (unmerge may include intermediate deletes that aren't present in the simple diff). My prediction is that eventually a user will be confused why there are deletes in their diff that are somewhat surprising, but it's obscure enough that I feel comfortable with only supporting unmerge for now if that's the preference.